### PR TITLE
Refactored groupPermissions

### PR DESCRIPTION
**Pull request by kurtisassad**
## Link to Issue
Closes: #TODO

## Description of Changes
- Adds FIXME widget to TODO page.
- 

## Test Plan
- Unit tested the `FIXME()` call.
- CA (click around) tested on local and frack:
  - TODO page
  - 

## Deployment Plan

1. 

## Other Considerations

-

### CODE REVIEW

File: packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+          },
+        );
+
+        for (const t of topics) {
+          for (const g of t.group_ids) {
+            // we have consistency issues with the old model.
+            // Some group_ids from topics are referencing non-existing groups, so skip them if we encounter them
+            if (g < 52) {
+              continue;
+            }
```
Comment (rbennettcw) [inline packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js:35]: Can we use a constant for this? Magic numbers don't give any context.

Comment (kurtisassad) [reply packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js:35]: Good point. I am going to use Tim's query to remove all the inconsistent topic_ids instead.

**Review by masvelio — APPROVED** [review]
FE code looks good to me

File: packages/commonwealth/server/controllers/server_groups_methods/update_group.ts [hunk 2]
```diff
@@ -118,53 +134,15 @@ export async function __updateGroup(
...
-            },
-          },
-          transaction,
-        },
+    // if associated with some topics, update the groupPermissions
+    if (topics) {
+      group.groupPermissions = await updateGroupPermissions(
+        topics,
+        allowedActions ??
+          (Object.values(ForumActionsEnum) as unknown as ForumActions),
```
Comment (timolegros) [inline packages/commonwealth/server/controllers/server_groups_methods/update_group.ts:142]: `ForumActions` type is a single enum property. This is setting an array of values to be equivalent to a single value. If the function should handle array values then the function type should be updated.

File: packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+          {
+            type: queryInterface.sequelize.QueryTypes.SELECT,
+            raw: true,
+            transaction,
+          },
+        );
+
+        for (const t of topics) {
+          for (const g of t.group_ids) {
+            await queryInterface.sequelize.query(
```
Comment (timolegros) [inline packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js:59]: `bulkInsert`?

File: packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            );
+          }
+        }
+      });
+    } catch (e) {
+      console.log('Migration failed');
+      throw e;
+    }
+
+    // For whatever reason this cannot be in the transaction, otherwise it will fail.
```
Comment (timolegros) [inline packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js:77]: ?????

File: packages/commonwealth/server/util/requirementsModule/validateTopicGroupsMembership.ts [hunk 5]
```diff
@@ -23,47 +23,34 @@ export async function validateTopicGroupsMembership(
...
     `
         SELECT g.*, gp.allowed_actions FROM "Groups" as g LEFT JOIN "GroupPermissions" gp ON g.id = gp.group_id
-        WHERE g.community_id = :communityId AND g.id IN(:groupIds);
+        WHERE g.community_id = :communityId AND gp.topic_id = :topicId;
       `,
     {
       type: QueryTypes.SELECT,
       raw: true,
-      replacements: { communityId, groupIds: topic.group_ids },
+      replacements: { communityId, topicId: topicId },
```
Comment (timolegros) [inline packages/commonwealth/server/util/requirementsModule/validateTopicGroupsMembership.ts:38]:
```suggestion
      replacements: { communityId, topicId },
```

**Review by timolegros — CHANGES_REQUESTED** [review]
Good work. Would like to remove the `unknown` type casts and optimize queries where possible. Will run through test plan after comments are handled.

Why is the column removal not included in the migration?

File: packages/commonwealth/client/scripts/models/Group.ts [hunk 6]
```diff
@@ -24,6 +24,7 @@ interface APIResponseFormat {
       token_id?: string;
     };
   }[];
+  GroupPermissions: { Topic: unknown }[];
```
Comment (mzparacha) [inline packages/commonwealth/client/scripts/models/Group.ts:27]: this doesn't actually get returned from `/groups`, why did we add this?

image

File: packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js (hunk 4 above)
Comment (kurtisassad) [reply packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js:77]: I mean there will be a deadlock and the migration will hang. Any solution or suggestion for this would be welcome

File: packages/commonwealth/server/controllers/server_groups_methods/update_group.ts (hunk 2 above)
Comment (kurtisassad) [reply packages/commonwealth/server/controllers/server_groups_methods/update_group.ts:142]: Good catch

File: packages/commonwealth/server/controllers/server_groups_methods/refresh_membership.ts [hunk 7]
```diff
@@ -13,63 +14,65 @@ const Errors = {
...
-    groups = groups.filter((g) => topic.group_ids.includes(g.id));
+  if (groups.length === 0 && topicId) {
+    throw new AppError(Errors.TopicNotFound);
   }
 
   const memberships = await refreshMembershipsForAddress(
     this.models,
     address,
-    groups,
+    groups as unknown as GroupAttributes[],
```
[the remark this answers is not in the document]
Comment (kurtisassad) [reply packages/commonwealth/server/controllers/server_groups_methods/refresh_membership.ts:58]: Another good catch

File: packages/commonwealth/client/scripts/models/Group.ts (hunk 6 above)
Comment (kurtisassad) [reply packages/commonwealth/client/scripts/models/Group.ts:27]: Good catch. Only included if some GroupPermissions exist for the group and includeTopics=true, otherwise returns undefined. I will update the model to make it optional.

File: packages/commonwealth/client/scripts/hooks/useForumActionGated.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    m.topicIds.map((t) => ({
+      topic_id: t,
+      allowedActions: m.allowedActions,
+    })),
+  ) as unknown as {
+    topic_id: number;
+    allowedActions: ForumActions[];
+  }[];
+
+  const topicIdToIsAllowedMap: Map<number, ForumActions[]> = new Map(
```
[the remark this answers is not in the document]
Comment (kurtisassad) [reply packages/commonwealth/client/scripts/hooks/useForumActionGated.ts:38]: Good idea

File: libs/model/src/community/CreateGroup.command.ts [hunk 9]
```diff
@@ -55,24 +56,20 @@ export function CreateGroup(): Command<typeof schemas.CreateGroup> {
...
-                  sequelize.col('group_ids'),
-                  group.id,
-                ),
-              },
-              {
-                where: {
-                  id: {
-                    [Op.in]: topicsToAssociate.map(({ id }) => id!),
-                  },
+            const createPermissionsPromises = topicsToAssociate.map((topic) => {
```
[the remark this answers is not in the document]
Comment (timolegros) [reply libs/model/src/community/CreateGroup.command.ts:59]: Ok so can we use type assertions or just raw SQL instead?

File: packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
+'use strict';
```
Comment (timolegros) [inline packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js:1]: re-order the migration to be the latest

File: packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js (hunk 4 above)
Comment (timolegros) [reply packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js:77]: Migration worked fine for me when the column drop is included in the migration. Is there a specific way to reproduce?

**Comment by kurtisassad** [comment]
> Found a bug with group updates

I am a little bit confused by this. So we first are creating a group with Horizon EON token. But then we are modifying it so that it must also be merfeld's 0x2Ab address. The issue is this address does not have any Horizon EON token, so it would not have both the token and be his address. That is why there are no members that qualify for the group. Maybe I am understanding it incorrectly?

TLDR: 0x2Ab does not have any Horizon EON token, so it is not have the requirements met to be part of the group.

I tried the same setup but also including Parsam's address (which also has Horizon EON token) and it worked.

File: libs/model/src/community/CreateGroup.command.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (kurtisassad) [reply libs/model/src/community/CreateGroup.command.ts:59]: Maybe we can use raw sql? But I need to make a prepared statement for each allowed_actions array. Is there an easy way to do this? Because the current way we use prepared statements is by specifying the individual substitution variables, but this is not possible to do in a loop I think.

File: packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js (hunk 1 above)
Comment (ForestMars) [reply packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js]: Resolved in [2328c70]

File: packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js (hunk 4 above)
Comment (kurtisassad) [reply packages/commonwealth/server/migrations/20240718135145-group-permission-refactor.js]: Hmm strange, ends up passing now instead of deadlocking. I am not sure how I produced it the first time, but I have made the change

File: libs/model/src/community/CreateGroup.command.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (timolegros) [reply libs/model/src/community/CreateGroup.command.ts:59]: `bulkCreate` seemed to work for me when I tried it but maybe I did something wrong. Either way, we should use a raw query if `bulkCreate` doesn't work. Not sure what you mean by a prepared statement. If you mean creating a loop to write the query + add the replacements/bindings then yes that is necessary and no there isn't really an easy way that I know of.

[the remark this answers is not in the document]
Comment (kurtisassad) [reply libs/model/src/community/CreateGroup.command.ts]: works with sequelize.literal

File: packages/commonwealth/test/unit/server_controllers/server_threads_controller.spec.ts [hunk 11]
```diff
@@ -345,140 +345,6 @@ describe('ServerThreadsController', () => {
         }),
       ).to.be.rejectedWith('Ban error: big ban err');
     });
-
-    test('should throw error (token balance)', () => {
```
Comment (kurtisassad) [inline packages/commonwealth/test/unit/server_controllers/server_threads_controller.spec.ts:349]: If anyone can figure out how we can make this test pass that would be great, otherwise I think it should be removed. Basically the stack trace it provides gives no insight.

Comment (rbennettcw) [reply packages/commonwealth/test/unit/server_controllers/server_threads_controller.spec.ts:349]: Yeah I think we might as well remove them if they're creating friction. With the upcoming TRPC refactors, they'll be obsolete anyway.

File: packages/commonwealth/server/controllers/server_groups_methods/update_group.ts [hunk 12]
```diff
@@ -177,3 +154,73 @@ export async function __updateGroup(
...
+  const upsertPromises = upsertGroupPermissions.map((p) =>
+    models.sequelize.query(
+      `
+            INSERT INTO "GroupPermissions" (group_id, topic_id, allowed_actions, created_at, updated_at) VALUES
+            (:groupId, :topicId, Array[:allowedActions]::"enum_GroupPermissions_allowed_actions"[], NOW(), NOW())
+            ON CONFLICT (group_id, topic_id) DO UPDATE
+            SET
+                allowed_actions = EXCLUDED.allowed_actions,
+                updated_at = EXCLUDED.updated_at
+            RETURNING *;
```
Comment (rbennettcw) [inline packages/commonwealth/server/controllers/server_groups_methods/update_group.ts:209]: We should always use a type for raw queries, e.g. `models.sequelize.query<GroupPermissionInstance>(...)`, then there's no need to cast the type later.

Also, I think wildcards `*` should be avoided. It's easier to write but tricker to debug due to lack of transparency.

File: packages/commonwealth/server/controllers/server_groups_methods/update_group.ts [hunk 13]
```diff
@@ -177,3 +154,73 @@ export async function __updateGroup(
...
+          allowedActions: p.allowed_actions,
+        },
+        transaction,
+      },
+    ),
+  );
+
+  return (await Promise.all(upsertPromises)) as unknown as Promise<
+    GroupPermissionInstance[]
+  >;
```
Comment (rbennettcw) [inline packages/commonwealth/server/controllers/server_groups_methods/update_group.ts:225]: No need to await if you're returning a promise. It'll resolve either way. Also, if setting the type on the query, this would simply become `return Promise.all(upsertPromises)`

**Review by rbennettcw — COMMENTED** [review]
Some small comments. Lmk when tests/conflicts are fixed and I'll test locally.

File: packages/commonwealth/server/controllers/server_groups_methods/update_group.ts (hunk 13 above)
Comment (timolegros) [reply packages/commonwealth/server/controllers/server_groups_methods/update_group.ts:225]: `return await` and `return Promise...` are not the same and I think `return await` is recommended now:
https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/errorhandling/returningpromises.md

File: packages/commonwealth/client/scripts/views/components/feed.tsx [hunk 14]
```diff
@@ -51,34 +51,23 @@ const FeedThread = ({ thread }: { thread: Thread }) => {
...
+    topicId: thread?.topic?.id,
   });
 
-  const isTopicGated = !!(memberships || []).find((membership) =>
-    membership.topicIds.includes(thread?.topic?.id),
+  const { canCreateComment, canReactToThread } = canPerformAction(
+    allowedActions,
+    isAdmin,
+    thread?.topic?.id,
   );
```
[the remark this answers is not in the document]
Comment (kurtisassad) [reply packages/commonwealth/client/scripts/views/components/feed.tsx:64]: Good catch, I will merge this into the useForumActionGated

**Comment by masvelio** [comment]
@kurtisassad any chance to make this PR mergable? Otherwise, let's close it to clean up the PR list
